txscript/pkscript.go

```diff
@@ -0,0 +1,194 @@
+	copy(outputScript.script[:], pkScript)
+
+	return outputScript, nil
+}
+
+// isSupportedScriptType determines whether the script type is supported by the
+// PkScript struct.
+func isSupportedScriptType(class ScriptClass) bool {
+	switch class {
+	case WitnessV0PubKeyHashTy, ScriptHashTy, WitnessV0ScriptHashTy:
+		return true
+	default:
+		return false
+	}
+}
+
+// Class returns the script type.
+func (s PkScript) Class() ScriptClass {
+	return s.class
+}
+
+// Script returns the script as a byte slice without any padding.
+func (s PkScript) Script() []byte {
+	switch s.class {
+	case WitnessV0PubKeyHashTy:
+		return s.script[:witnessV0PubKeyHashLen]
+	case ScriptHashTy:
+		return s.script[:scriptHashLen]
+	case WitnessV0ScriptHashTy:
+		return s.script[:witnessV0ScriptHashLen]
+	default:
+		// Unsupported script type.
+		return nil
+	}
+}
+
+// Address encodes the script into an address for the given chain.
+func (s PkScript) Address(chainParams *chaincfg.Params) (btcutil.Address, error) {
+	_, addrs, _, err := ExtractPkScriptAddrs(s.Script(), chainParams)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse address: %v", err)
+	}
+
+	return addrs[0], nil
+}
+
+// String returns a hex-encoded string representation of the script.
+func (s PkScript) String() string {
+	return fmt.Sprintf("%x", s.Script())
+}
+
+// ComputePkScript computes the pkScript of an transaction output by looking at
+// the transaction input's signature script or witness.
+//
+// NOTE: Only P2SH, P2WSH, and P2WPKH redeem scripts are supported.
+func ComputePkScript(sigScript []byte, witness wire.TxWitness) (PkScript, error) {
+	var pkScript PkScript
+
+	// If a signature script is provided, since we only support P2SH for
+	// non-segwit scripts, then we'll assume this is a P2SH redeem script
+	// based on its length not being that of a P2PKH redeem script.
+	if len(sigScript) > 0 && len(sigScript) != pubKeyHashSigScriptLen {
+		scriptHash := hash160(sigScript)
+		script, err := payToScriptHashScript(scriptHash)
+		if err != nil {
+			return pkScript, err
+		}
+
+		pkScript.class = ScriptHashTy
+		copy(pkScript.script[:], script)
+
+		return pkScript, nil
+	}
+
+	// Otherwise, this is the redeem script for an unsupported non-segwit
+	// script type.
+	if len(sigScript) > 0 || len(witness) == 0 {
+		return pkScript, ErrUnsupportedScriptType
+	}
+
+	// We'll look at the witness stack to determine the proper pkScript
+	// type.
+	lastWitnessItem := witness[len(witness)-1]
+
+	switch {
+	// If the witness stack has a size of 2 and its last item is a
+	// compressed public key, then this is a P2WPKH witness.
+	case len(witness) == 2 && len(lastWitnessItem) == 33:
+		pubKeyHash := hash160(lastWitnessItem)
+		script, err := payToWitnessPubKeyHashScript(pubKeyHash)
+		if err != nil {
+			return pkScript, err
+		}
+
+		pkScript.class = WitnessV0PubKeyHashTy
+		copy(pkScript.script[:], script)
+
+	// For any other witnesses, we'll assume it's a P2WSH witness.
```

Comment:
Rather than just assuming this, we should SHA2 the last item and check that it produces the same hash as the raw witness program. If it doesn't , then we return an error as it's unknown.

Reply:
Doesn't seem like this can be done without having the original output's script, which is what we're attempting to compute.